src/signTxCatalystRegistration.c

```diff
@@ -0,0 +1,576 @@
+	VALIDATE(subctx->state == expected, ERR_INVALID_STATE);
+}
+
+static inline void advanceState()
+{
+	TRACE("Advancing Catalyst registration state from: %d", subctx->state);
+
+	switch (subctx->state) {
+
+	case STATE_CATALYST_REGISTRATION_VOTING_KEY:
+		subctx->state = STATE_CATALYST_REGISTRATION_STAKING_KEY;
+		break;
+
+	case STATE_CATALYST_REGISTRATION_STAKING_KEY:
+		subctx->state = STATE_CATALYST_REGISTRATION_VOTING_REWARDS_ADDRESS;
+		break;
+
+	case STATE_CATALYST_REGISTRATION_VOTING_REWARDS_ADDRESS:
+		subctx->state = STATE_CATALYST_REGISTRATION_NONCE;
+		break;
+
+	case STATE_CATALYST_REGISTRATION_NONCE:
+		subctx->state = STATE_CATALYST_REGISTRATION_CONFIRM;
+		break;
+
+	case STATE_CATALYST_REGISTRATION_CONFIRM:
+		subctx->state = STATE_CATALYST_REGISTRATION_FINISHED;
+		break;
+
+	default:
+		ASSERT(false);
+	}
+
+	TRACE("Advancing Catalyst registration state to: %d", subctx->state);
+}
+
+// ============================== VOTING KEY ==============================
+
+enum {
+	HANDLE_VOTING_KEY_STEP_DISPLAY = 8200,
+	HANDLE_VOTING_KEY_STEP_RESPOND,
+	HANDLE_VOTING_KEY_STEP_INVALID,
+};
+
+static void signTxCatalystRegistration_handleVotingKey_ui_runStep()
+{
+	TRACE("UI step %d", subctx->ui_step);
+	TRACE_STACK_USAGE();
+	ui_callback_fn_t* this_fn = signTxCatalystRegistration_handleVotingKey_ui_runStep;
+
+	UI_STEP_BEGIN(subctx->ui_step);
+
+	UI_STEP(HANDLE_VOTING_KEY_STEP_DISPLAY) {
+		STATIC_ASSERT(SIZEOF(subctx->stateData.votingPubKey) == CATALYST_VOTING_PUBLIC_KEY_LENGTH, "wrong voting public key size");
+
+		// Jormungandr public key, hence the "ed25519_pk" prefix
+		// https://github.com/input-output-hk/jormungandr/blob/a057af27493d823be02480bb20258c25ff979e2a/jormungandr-lib/src/crypto/key.rs#L126
+		ui_displayBech32Screen(
+		        "Voting public key",
+		        "ed25519_pk",
+		        subctx->stateData.votingPubKey, CATALYST_VOTING_PUBLIC_KEY_LENGTH,
+		        this_fn
+		);
+	}
+	UI_STEP(HANDLE_VOTING_KEY_STEP_RESPOND) {
+		respondSuccessEmptyMsg();
+		advanceState();
+	}
+	UI_STEP_END(HANDLE_VOTING_KEY_STEP_INVALID);
+}
+
+__noinline_due_to_stack__
+static void signTxCatalystRegistration_handleVotingKeyAPDU(uint8_t* wireDataBuffer, size_t wireDataSize)
+{
+	{
+		CHECK_STATE(STATE_CATALYST_REGISTRATION_VOTING_KEY);
+
+		ASSERT(wireDataSize < BUFFER_SIZE_PARANOIA);
+	}
+	{
+		explicit_bzero(&subctx->stateData, SIZEOF(subctx->stateData));
+	}
+	{
+		TRACE_BUFFER(wireDataBuffer, wireDataSize);
+
+		VALIDATE(wireDataSize == SIZEOF(subctx->stateData.votingPubKey), ERR_INVALID_DATA);
+
+		{
+			STATIC_ASSERT(SIZEOF(subctx->stateData.votingPubKey) == CATALYST_VOTING_PUBLIC_KEY_LENGTH, "wrong voting public key size");
+			os_memmove(subctx->stateData.votingPubKey, wireDataBuffer, CATALYST_VOTING_PUBLIC_KEY_LENGTH);
+		}
+	}
+	{
+		auxDataHashBuilder_catalystRegistration_enter(auxDataHashBuilder);
+		auxDataHashBuilder_catalystRegistration_enterPayload(auxDataHashBuilder);
```

Comment:
I would consider moving such things into `advanceState`. That's how we do it in signTx.c (handle*APDU only passes currently received data to txhashbuilder).

Reply:
The advanceState function is called after the first APDU that already handled the voting key and is supposed to add it to the hash context. It could be added to advanceState function in the metadata step in signTx but this would IMHO leak responsibilities from the submachine. Another solution could be making that part of the initialization of the aux data hash builder, but that also doesn't seem OK as the aux data hash builder is meant potentially for any metadata, not just a catalyst registration, so this is the most appropriate place I found after all.

Do you see how to introduce those "enter steps" sensibly into advanceState?